utils/formatting/encoding.go

```diff
@@ -4,51 +4,141 @@
 package formatting
 
 import (
+	"bytes"
+	"encoding/hex"
+	"errors"
 	"fmt"
+	"math"
+	"strings"
+
+	"github.com/ava-labs/avalanchego/utils/hashing"
+	"github.com/mr-tron/base58/base58"
 )
 
 const (
-	// HexEncoding specifies a hex plus 4 byte checksum encoding format
-	HexEncoding = "hex"
-	// CB58Encoding specifies the CB58 encoding format
-	CB58Encoding = "cb58"
+	// maximum length byte slice can be marshalled to a string
+	// using the CB58 encoding. Must be longer than the length
+	// of an ID and longer than the length of a SECP256k1 private key
+	// TODO: Reduce to a reasonable amount (e.g. 16 * 1024) after we
+	// give users a chance to export very large keystore users to hex
+	maxCB58Size = math.MaxInt32
+
+	checksumLen = 4
+)
+
+var (
+	errInvalidEncoding  = errors.New("invalid encoding")
+	errMissingChecksum  = errors.New("input string is smaller than the checksum size")
+	errBadChecksum      = errors.New("invalid input checksum")
+	errMissingHexPrefix = errors.New("missing 0x prefix to hex encoding")
 )
 
-// Encoding returns a struct used to format bytes for a specific encoding
-type Encoding interface {
-	ConvertBytes([]byte) string
-	ConvertString(str string) ([]byte, error)
-	Encoding() string
+// Encoding defines how bytes are converted to a string and vice versa
+type Encoding uint8
+
+const (
+	// Hex specifies a hex plus 4 byte checksum encoding format
+	Hex Encoding = iota
+	// CB58 specifies the CB58 encoding format
+	CB58
```

Comment:
CB58 needs to be the zero value, otherwise this PR will change the default encoding for all our APIs.